src/outputtype.cpp

```diff
@@ -82,11 +82,11 @@ std::vector<CTxDestination> GetAllDestinationsForKey(const CPubKey& key)
     }
 }
 
-CTxDestination AddAndGetDestinationForScript(FillableSigningProvider& keystore, const CScript& script, OutputType type)
+CTxDestination AddAndGetDestinationForScript(FlatSigningProvider& keystore, const CScript& script, OutputType type)
 {
     // Add script to keystore
-    keystore.AddCScript(script);
-    // Note that scripts over 520 bytes are not yet supported.
+    keystore.scripts.emplace(CScriptID(script), script);
```

Comment:
Just curious, why doesn't `FlatSigningProvider` have a `AddCScript()` method like it's Fillable counterpart? I was surprised to see the maps in the provider aren't private or protected, I thought that was more our style.

Reply:
> Just curious, why doesn't FlatSigningProvider have a AddCScript() method like it's Fillable counterpart? I was surprised to see the maps in the provider aren't private or protected, I thought that was more our style.

`FillableSigningProvider` is the legacy counterpart. All other `SigningProvider` subclasses were introduced with the output descriptors implementation.

If you look at the new providers, all of them are loaded in place and stay constant over time. They are only accessed through the `SigningProvider` interface methods and not directly (this interface only present getters). In contraposition, the legacy provider fulfills both roles. It implements the `SigningProvider` methods and, at the same time, allows new scripts and keys to be added over time.
This reflects the legacy vs descriptors difference: each new import in the descriptors wallet refers to a new spkm, while in the legacy wallet, it  refers to a new script/key added into the single legacy spkm.

Reply:
So all the places where we call `FlatSigningProvider.scripts.emplace()` doesn't count as "allows new scripts and keys to be added over time" ?

Reply:
> So all the places where we call `FlatSigningProvider.scripts.emplace()` doesn't count as "allows new scripts and keys to be added over time" ?

As uses to be in software development, the answer is "it depends".
As far as I can tell, the idea was to hide the `SigningProvider` implementation. Which is sound.

I haven't looked at all the places where the `FlatSigningProvider` is used but most of them should be inside the `descriptor.cpp` and similar related processes where the parent process calls a function that loads the provider with some keys/scripts, to then use them and discard the provider.

When I said "allows new scripts and keys to be added over time" was referring to keep it in-memory. And continually add new elements to it. Which should not be happening with `FlatSigningProvider` anywhere.